libs/wingc/src/fold.rs

```diff
@@ -0,0 +1,438 @@
+			variable: f.fold_reference(variable),
+			value: f.fold_expr(value),
+		},
+		StmtKind::Scope(scope) => StmtKind::Scope(f.fold_scope(scope)),
+		StmtKind::Class(class) => StmtKind::Class(f.fold_class(class)),
+		StmtKind::Interface(interface) => StmtKind::Interface(f.fold_interface(interface)),
+		StmtKind::Struct { name, extends, fields } => StmtKind::Struct {
+			name: f.fold_symbol(name),
+			extends: extends.into_iter().map(|e| f.fold_user_defined_type(e)).collect(),
+			fields: fields.into_iter().map(|field| f.fold_class_field(field)).collect(),
+		},
+		StmtKind::Enum { name, values } => StmtKind::Enum {
+			name: f.fold_symbol(name),
+			values: values.into_iter().map(|value| f.fold_symbol(value)).collect(),
+		},
+		StmtKind::TryCatch {
+			try_statements,
+			catch_block,
+			finally_statements,
+		} => StmtKind::TryCatch {
+			try_statements: f.fold_scope(try_statements),
+			catch_block: catch_block.map(|catch_block| CatchBlock {
+				exception_var: catch_block.exception_var.map(|var| f.fold_symbol(var)),
+				statements: f.fold_scope(catch_block.statements),
+			}),
+			finally_statements: finally_statements.map(|statements| f.fold_scope(statements)),
+		},
+	};
+	Stmt {
+		kind,
+		span: node.span,
+		idx: node.idx,
+	}
+}
+
+pub fn fold_class<F>(f: &mut F, node: Class) -> Class
+where
+	F: Fold + ?Sized,
+{
+	Class {
+		name: f.fold_symbol(node.name),
+		fields: node.fields.into_iter().map(|field| f.fold_class_field(field)).collect(),
+		methods: node
+			.methods
+			.into_iter()
+			.map(|(name, def)| (f.fold_symbol(name), f.fold_function_definition(def)))
+			.collect(),
+		constructor: f.fold_constructor(node.constructor),
+		parent: node.parent.map(|parent| f.fold_user_defined_type(parent)),
+		implements: node
+			.implements
+			.into_iter()
+			.map(|interface| f.fold_user_defined_type(interface))
+			.collect(),
+		is_resource: node.is_resource,
+	}
+}
+
+pub fn fold_class_field<F>(f: &mut F, node: ClassField) -> ClassField
+where
+	F: Fold + ?Sized,
+{
+	ClassField {
+		name: f.fold_symbol(node.name),
+		member_type: f.fold_type_annotation(node.member_type),
+		reassignable: node.reassignable,
+		phase: node.phase,
+		is_static: node.is_static,
+	}
+}
+
+pub fn fold_interface<F>(f: &mut F, node: Interface) -> Interface
+where
+	F: Fold + ?Sized,
+{
+	Interface {
+		name: f.fold_symbol(node.name),
+		methods: node
+			.methods
+			.into_iter()
+			.map(|(name, sig)| (f.fold_symbol(name), f.fold_function_signature(sig)))
+			.collect(),
+		extends: node
+			.extends
+			.into_iter()
+			.map(|interface| f.fold_user_defined_type(interface))
+			.collect(),
+	}
+}
+
+pub fn fold_constructor<F>(f: &mut F, node: Constructor) -> Constructor
+where
+	F: Fold + ?Sized,
+{
+	Constructor {
+		signature: f.fold_function_signature(node.signature),
+		statements: f.fold_scope(node.statements),
+	}
+}
+
+pub fn fold_expr<F>(f: &mut F, node: Expr) -> Expr
+where
+	F: Fold + ?Sized,
+{
+	let kind = match node.kind {
```

Comment:
I am a bit concerned about maintaining this... Do we have a way to enforce that when we update the model it will requiring updating this code?

I guess Rust algebraic types help...

Reply:
yeah, Rust's type help a lot, but it's still possible to miss things. It's actually a bigger issue for `visit` since it's possible for someone to change e.g. add a new field to `ast::Class`, and forget to update the `Visit` trait so that the field is recursively visited. Ideally we would like to safely code-generate these traits with Rust's macro system.